src/libZilliqa/Zilliqa.cpp

```diff
@@ -201,6 +189,30 @@ Zilliqa::Zilliqa(const PairOfKey& key, const Peer& peer, SyncType syncType,
     }
   }
 
+  // when individual node is being recovered and persistence is not available
+  // locally, then Rejoin as if new miner node which will download persistence
+  // from S3 incremental db and identify if already part of any
+  // shard/dscommittee and proceed accordingly.
+
+  if (!LOOKUP_NODE_MODE && (SyncType::RECOVERY_ALL_SYNC == syncType)) {
+    if (!boost::filesystem::exists(STORAGE_PATH + PERSISTENCE_PATH)) {
+      syncType = SyncType::NEW_SYNC;
+      m_lookup.SetSyncType(SyncType::NEW_SYNC);
+    }
+  }
+
+  if (SyncType::NEW_SYNC == syncType) {
+    // Setting it earliest before even p2pcomm is instantiated
+    m_n.m_runFromLate = true;
+  }
+
+  P2PComm::GetInstance().SetSelfPeer(peer);
```

Comment:
Will `IsNodeInDSGuardList` and `IsNodeInShardGuardList` above work if `SetSelfPeer` and `SetSelfKey` are now set later?

Reply:
ohh ya..good catch. when tested against guard nodes it works. But it won't log correctly.